lib/Alert/Alert.man
'\" t
.TH "" 3 "" "Version 3.0" "Free Widget Foundation"
.SH NAME
XfwfAlert
.SH DESCRIPTION
The XfwfAlert widget is a pop-up dialog box with three parts: an
icon in the top left corner, a text area, and a row of buttons along
the bottom. Usually the alert widget is used for error messages,
warnings and other kinds of informational messages. The icon indicates
the type of message: (fatal) error, warning, other message, request
for confirmation, etc. There are some built-in icons for these types.
Most alerts will have one or two buttons, e.g., OK and CANCEL.

An alert widget can be created in the usual way for shells, with
\fIXtVaCreatePopup\fP and similar routines, but a much more convenient way
is a call to \fIXfwfDialog\fP. \fIXfwfDialog\fP creates a temporary alert
widget and waits for the user to acknowledge it by clicking one of the
buttons. The function returns the number of the button that was
clicked.

An alert is made up of XfwfBoard(3X11), XfwfIcon(3X11),
Xfwflabel(3X11) and XfwfButton(3X11) widgets. The label widget will be
set to the minimum size that will show the full message. The alert
widget will be as small as possible to show all its children.

.SS "Public variables"

.ps -2
.TS
center box;
cBsss
lB|lB|lB|lB
l|l|l|l.
XfwfAlert
Name	Class	Type	Default
XtNicon	XtCIcon	Icon *	"NONE"
XtNmessage	XtCMessage	String 	NULL 
XtNbuttons	XtCButtons	String 	NULL 
XtNlastChoice	XtCLastChoice	int 	0 
XtNcallback	XtCCallback	Callback	NULL 

.TE
.ps +2

.TP
.I "XtNicon"
The icon is of type \fIIcon\fP (see XfwfIcon(3X11)). The converter can load
images from file and it knows about 6 built-in icons: \fI"FATAL"\fP, \fI"ERROR"\fP,
\fI"WARNING"\fP, \fI"QUESTION"\fP, \fI"INFO"\fP and \fI"NONE"\fP. The \fINONE\fP icon is an
invisible (i.e., transparent) image of the same size as the others.

.TP
.I "XtNmessage"
The message is a simple string, that may include newlines.

.TP
.I "XtNbuttons"
The buttons are created with labels which will be taken from the
\fIbuttons\fP resource. \fIbuttons\fP is a string with the labels separated by
vertical bars. If the string is empty, a single button with label
``Ok'' will be created.

.TP
.I "XtNlastChoice"
The last button that was selected is also the default for the next
time (as soon as keyboard traversal has been implemented\dots) Buttons
are numbered starting with 0. Set this resource to -1 if you don't
want a default button.

.TP
.I "XtNcallback"
When a button is clicked, the number of the button is stored in
\fIlastChoice\fP, before the \fIcallback\fP functions are called. The
callbacks is passed the number in the \fIcall_data\fP argument. The
buttons are numbered starting at 0.

.ps -2
.TS
center box;
cBsss
lB|lB|lB|lB
l|l|l|l.
TransientShell
Name	Class	Type	Default
transientFor	XtCTransientFor	Widget 	NULL 

.TE
.ps +2

.ps -2
.TS
center box;
cBsss
lB|lB|lB|lB
l|l|l|l.
WMShell
Name	Class	Type	Default
baseHeight	XtCBaseHeight	Int 	XtUnspecifiedShellInt 
baseWidth	XtCBaseWidth	Int 	XtUnspecifiedShellInt 
XtNtitle	XtCTitle	String 	NULL 
titleEncoding	XtCTitleEncoding	Atom 	XA_STRING 
XtNtransient	XtCTransient	Boolean 	False 
waitForWm	XtCWaitForWm	Boolean 	True 
winGravity	XtCWinGravity	Int 	NULL 
wmTimeout	XtCWmTimeout	Int 	5000 

.TE
.ps +2

.ps -2
.TS
center box;
cBsss
lB|lB|lB|lB
l|l|l|l.
Shell
Name	Class	Type	Default
allowShellResize	XtCAllowShellResize	Boolean 	False 
createPopupChildProc	XtCCreatePopupChildProc	XTCreatePopupChildProc 	NULL 
XtNgeometry	XtCGeometry	String 	NULL 
overrideRedirect	XtCOverrideRedirect	Boolean 	False 
popdownCallback	XtCPopdownCallback	Callback	NULL 
popupCallback	XtCPopupCallback	Callback	NULL 
saveUnder	XtCSaveUnder	Boolean 	False 
XtNvisual	XtCVisual	Visual *	CopyFromParent 

.TE
.ps +2

.ps -2
.TS
center box;
cBsss
lB|lB|lB|lB
l|l|l|l.
Composite
Name	Class	Type	Default
XtNchildren	XtCChildren	WidgetList 	NULL 
insertPosition	XtCInsertPosition	XTOrderProc 	NULL 
numChildren	XtCNumChildren	Cardinal 	0 

.TE
.ps +2

.ps -2
.TS
center box;
cBsss
lB|lB|lB|lB
l|l|l|l.
Core
Name	Class	Type	Default
XtNx	XtCX	Position 	0 
XtNy	XtCY	Position 	0 
XtNwidth	XtCWidth	Dimension 	0 
XtNheight	XtCHeight	Dimension 	0 
borderWidth	XtCBorderWidth	Dimension 	0 
XtNcolormap	XtCColormap	Colormap 	NULL 
XtNdepth	XtCDepth	Int 	0 
destroyCallback	XtCDestroyCallback	XTCallbackList 	NULL 
XtNsensitive	XtCSensitive	Boolean 	True 
XtNtm	XtCTm	XTTMRec 	NULL 
ancestorSensitive	XtCAncestorSensitive	Boolean 	False 
accelerators	XtCAccelerators	XTTranslations 	NULL 
borderColor	XtCBorderColor	Pixel 	0 
borderPixmap	XtCBorderPixmap	Pixmap 	NULL 
background	XtCBackground	Pixel 	0 
backgroundPixmap	XtCBackgroundPixmap	Pixmap 	NULL 
mappedWhenManaged	XtCMappedWhenManaged	Boolean 	True 
XtNscreen	XtCScreen	Screen *	NULL 

.TE
.ps +2

.SS "Exports"

As stated above, the alert widget can be created in the normal way
as a Shell widget with a particular set of resources, but a more
convenient way in many cases will be the \fIXfwfDialog\fP function. The
function creates an alert box with an icon, message and buttons, pops
it up, and returns only when the user has clicked one of the buttons.
The function's result will be the number of the button that was
clicked.

The first argument, \fItoplevel\fP, will be the widget that is used as the
alert's parent. Usually that will be the top level window of the
application, but it need not be. The alert box will appear
approximately in the middle of the \fItoplevel\fP widget (if the window
manager allows it).

The second argument is the name of the icon that will be shown in the
upper left corner of the alert. It can be one of the built-in names
\fI"FATAL"\fP, \fI"ERROR"\fP, \fI"WARNING"\fP, \fI"QUESTION"\fP, \fI"INFO"\fP and
\fI"NONE"\fP; or it can be the name of a file that contains an icon in XPM
format. The icon widget is of class XfwfIcon(3X11). It is named
\fI"_icon"\fP.

The third argument is the message to display in the center of the
alert. It may contain newlines. The message will be put into an
XfwfLabel(3X11) widget, named, appropriately, \fI"_message"\fP.

The last argument, \fIbuttons\fP, is a single string containing the labels
for all buttons, separated with a vertical bar. If this argument is
\fINULL\fP, that doesn't mean that there will be no buttons. Instead, a
single button with the label \fI"OK"\fP will be created. After all, if
there are no buttons, how will the alert box ever disappear? The
buttons are of type XfwfButton(3X11) and are named \fI"1"\fP, \fI"2"\fP, etc.

The alert itself will simply be called \fI"alert"\fP and its child, an
XfwfBoard(3X11) widget, will be called \fI"_board"\fP.

The implementation is as follows: After the alert has been created,
\fIXfwfDialog\fP adds a callback to the alert widget, which will set the
\fIchoice\fP variable when a button is pressed. The function then enters a
special purpose event loop, which dispatches events normally, but only
until the \fIchoice\fP variable is changed. The function returns with the
value of \fIchoice\fP

\fBdef\fP max(a, b) =
((a )>(b )?(a ):(b ))

\fBdef\fP min(a, b) =
((a )<(b )?(a ):(b ))

.nf
int  XfwfDialog(String  name, Widget  parent, String  icon, String  msg, String  buttons)
.fi



debian/changelog
fwf (4.0.1-1) unstable; urgency=low

  * Build using autoconf/automake.
  * Compile for x86 and amd64 processors.

 -- Erling Andersen <Erling.Andersen@Helse-Bergen.NO>  Mon, 09 Feb 2009 10:23:59 +0100


lib/Alert/Alert.c
/* Generated by wbuild from "Alert.w"
** (generator version $Revision$ of $Date$)
*/
#include <X11/IntrinsicP.h>
#include <X11/StringDefs.h>
#line 348 "Alert.w"
#include <stdlib.h>
#line 349 "Alert.w"
#include <stdio.h>
#line 350 "Alert.w"
#include <../Icon/Icon.h>
#line 351 "Alert.w"
#include <../Label/Label.h>
#line 352 "Alert.w"
#include <../RowCol/RowCol.h>
#line 353 "Alert.w"
#include <../Button/Button.h>
#line 354 "Alert.w"
#include <../converters/Converters.h>
#include <AlertP.h>
static void _resolve_inheritance(
#if NeedFunctionPrototypes
WidgetClass
#endif
);
#line 290 "Alert.w"
static void class_initialize(
#if NeedFunctionPrototypes
void
#endif
);
#line 303 "Alert.w"
static void initialize(
#if NeedFunctionPrototypes
Widget ,Widget,ArgList ,Cardinal *
#endif
);
#line 319 "Alert.w"
static Boolean  set_values(
#if NeedFunctionPrototypes
Widget ,Widget ,Widget,ArgList ,Cardinal *
#endif
);
#line 339 "Alert.w"
static void destroy(
#if NeedFunctionPrototypes
Widget
#endif
);
#line 156 "Alert.w"
#define ALLOC 1024 


#line 157 "Alert.w"
#define myrealloc(p, size) (void *)XtRealloc ((void *)p ,(size +ALLOC -1 )/ALLOC *ALLOC )


#line 159 "Alert.w"
#define mymalloc(size) (void *)XtMalloc (size )


#line 160 "Alert.w"
#define myfree(p) XtFree ((void *)p )


#line 167 "Alert.w"
static void button_pressed(
#if NeedFunctionPrototypes
Widget ,XtPointer ,XtPointer 
#endif
);
#line 176 "Alert.w"
static void click(
#if NeedFunctionPrototypes
Widget ,XtPointer ,XtPointer 
#endif
);
#line 188 "Alert.w"
static void create_children(
#if NeedFunctionPrototypes
Widget
#endif
);
#line 167 "Alert.w"
/*ARGSUSED*/
#if NeedFunctionPrototypes
#line 167 "Alert.w"
static void button_pressed(Widget  w,XtPointer  client_data,XtPointer  call_data)
#else
#line 167 "Alert.w"
static void button_pressed(w,client_data,call_data)Widget  w;XtPointer  client_data;XtPointer  call_data;
#endif
#line 168 "Alert.w"
{
    XtPopdown(w);
    *((int*) client_data) = (int) call_data;
}
#line 176 "Alert.w"
/*ARGSUSED*/
#if NeedFunctionPrototypes
#line 176 "Alert.w"
static void click(Widget  w,XtPointer  client_data,XtPointer  call_data)
#else
#line 176 "Alert.w"
static void click(w,client_data,call_data)Widget  w;XtPointer  client_data;XtPointer  call_data;
#endif
#line 177 "Alert.w"
{
    Widget self = XtParent(XtParent(w));

    ((XfwfAlertWidget)self)->xfwfAlert.lastChoice = (int) client_data;
    XtCallCallbackList(self, ((XfwfAlertWidget)self)->xfwfAlert.callback, (XtPointer) ((XfwfAlertWidget)self)->xfwfAlert.lastChoice);
}
#line 188 "Alert.w"
/*ARGSUSED*/
#if NeedFunctionPrototypes
#line 188 "Alert.w"
static void create_children(Widget self)
#else
#line 188 "Alert.w"
static void create_children(self)Widget self;
#endif
#line 189 "Alert.w"
{
    Dimension ht, maxht, iconwd, iconht, labelwd, wd, bframe, labelht;
    Widget toprow, bottomrow, iconw, labelw, btn;
    Position x, buttony;
    String buttons, p;
    char name[20];
    int i;

    /*
     * Destroy old copies of the children
     */
    if (((XfwfAlertWidget)self)->xfwfAlert.iconw) XtDestroyWidget(((XfwfAlertWidget)self)->xfwfAlert.iconw);
    if (((XfwfAlertWidget)self)->xfwfAlert.labelw) XtDestroyWidget(((XfwfAlertWidget)self)->xfwfAlert.labelw);
    for (i = 0; i < ((XfwfAlertWidget)self)->xfwfAlert.num_buttons; i++) XtDestroyWidget(((XfwfAlertWidget)self)->xfwfAlert.buttonw[i]);

    bframe = XfwfCallFrameWidth(((XfwfAlertWidget)self)->xfwfAlert.outer);	/* Board's frame */
    if (((XfwfAlertWidget)self)->xfwfAlert.buttons) buttons = XtNewString(((XfwfAlertWidget)self)->xfwfAlert.buttons); /* Temporary copy */

    iconw = XtVaCreateManagedWidget
	("icon", xfwfIconWidgetClass, ((XfwfAlertWidget)self)->xfwfAlert.outer,
	 XtNx, bframe,
	 XtNy, bframe,
	 XtNimage, ((XfwfAlertWidget)self)->xfwfAlert.icon,
	 XtNtraversalOn, False,
	 NULL);
    iconwd = ((XfwfAlertWidget)iconw)->core.width;
    iconht = ((XfwfAlertWidget)iconw)->core.height;

    labelw = XtVaCreateManagedWidget
	("label", xfwfLabelWidgetClass, ((XfwfAlertWidget)self)->xfwfAlert.outer,
	 XtNx, bframe + iconwd,
	 XtNy, bframe,
	 XtNshrinkToFit, True,
	 XtNlabel, ((XfwfAlertWidget)self)->xfwfAlert.message,
	 NULL);
    labelwd = ((XfwfAlertWidget)labelw)->core.width;
    labelht = ((XfwfAlertWidget)labelw)->core.height;

    buttony = bframe + max(iconht, labelht);
    x = bframe + iconwd;
    maxht = 0;
    ((XfwfAlertWidget)self)->xfwfAlert.num_buttons = 0;
    if (((XfwfAlertWidget)self)->xfwfAlert.buttons) {
	i = 0;
	p = strtok(buttons, "|");
	while (p) {
	    ((XfwfAlertWidget)self)->xfwfAlert.buttonw = myrealloc(((XfwfAlertWidget)self)->xfwfAlert.buttonw, (i + 1) * sizeof(*((XfwfAlertWidget)self)->xfwfAlert.buttonw));
	    sprintf(name, "%d", i);
	    btn = XtVaCreateManagedWidget
		(name, xfwfButtonWidgetClass, ((XfwfAlertWidget)self)->xfwfAlert.outer,
		 XtNx, x,
		 XtNy, buttony,
		 XtNshrinkToFit, True,
		 XtNlabel, p,
		 NULL);
	    wd = ((XfwfAlertWidget)btn)->core.width;
	    ht = ((XfwfAlertWidget)btn)->core.height;
	    XtAddCallback(btn, XtNactivate, click, (XtPointer) i);
	    maxht = max(maxht, ht);
	    x += wd;
	    ((XfwfAlertWidget)self)->xfwfAlert.buttonw[i] = btn;
	    i++;
	    p = strtok(NULL, "|");
	    ((XfwfAlertWidget)self)->xfwfAlert.num_buttons++;
	}
	((XfwfAlertWidget)self)->xfwfAlert.num_buttons = i;
    } else {
	((XfwfAlertWidget)self)->xfwfAlert.buttonw = mymalloc(2 * sizeof(*((XfwfAlertWidget)self)->xfwfAlert.buttonw));
	((XfwfAlertWidget)self)->xfwfAlert.num_buttons = 1;
	btn = XtVaCreateManagedWidget
	    ("0", xfwfButtonWidgetClass, ((XfwfAlertWidget)self)->xfwfAlert.outer,
	     XtNx, x,
	     XtNy, buttony,
	     XtNshrinkToFit, True,
	     XtNlabel, " OK ",
	     NULL);
	wd = ((XfwfAlertWidget)btn)->core.width;
	ht = ((XfwfAlertWidget)btn)->core.height;
	XtAddCallback(btn, XtNactivate, click, 0);
	maxht = ht;
	x += wd;
	((XfwfAlertWidget)self)->xfwfAlert.buttonw[0] = btn;
    }
    if (((XfwfAlertWidget)self)->xfwfAlert.buttons) XtFree(buttons);

    /*
     * And finally resize the outer Board
     */
    XtResizeWidget(((XfwfAlertWidget)self)->xfwfAlert.outer, max(x + bframe, iconwd + labelwd + 2 * bframe),
		   buttony + maxht + bframe, 0);

    ((XfwfAlertWidget)self)->xfwfAlert.iconw = iconw;
    ((XfwfAlertWidget)self)->xfwfAlert.labelw = labelw;
}

static XtResource resources[] = {
#line 35 "Alert.w"
{XtNicon,XtCIcon,XtRIcon,sizeof(((XfwfAlertRec*)NULL)->xfwfAlert.icon),XtOffsetOf(XfwfAlertRec,xfwfAlert.icon),XtRString,(XtPointer)"NONE"},
#line 39 "Alert.w"
{XtNmessage,XtCMessage,XtRString,sizeof(((XfwfAlertRec*)NULL)->xfwfAlert.message),XtOffsetOf(XfwfAlertRec,xfwfAlert.message),XtRImmediate,(XtPointer)NULL },
#line 46 "Alert.w"
{XtNbuttons,XtCButtons,XtRString,sizeof(((XfwfAlertRec*)NULL)->xfwfAlert.buttons),XtOffsetOf(XfwfAlertRec,xfwfAlert.buttons),XtRImmediate,(XtPointer)NULL },
#line 53 "Alert.w"
{XtNlastChoice,XtCLastChoice,XtRInt,sizeof(((XfwfAlertRec*)NULL)->xfwfAlert.lastChoice),XtOffsetOf(XfwfAlertRec,xfwfAlert.lastChoice),XtRImmediate,(XtPointer)0 },
#line 60 "Alert.w"
{XtNcallback,XtCCallback,XtRCallback,sizeof(((XfwfAlertRec*)NULL)->xfwfAlert.callback),XtOffsetOf(XfwfAlertRec,xfwfAlert.callback),XtRImmediate,(XtPointer)NULL },
};

XfwfAlertClassRec xfwfAlertClassRec = {
{ /* core_class part */
/* superclass   	*/  (WidgetClass) &transientShellClassRec,
/* class_name   	*/  "XfwfAlert",
/* widget_size  	*/  sizeof(XfwfAlertRec),
/* class_initialize 	*/  class_initialize,
/* class_part_initialize*/  _resolve_inheritance,
/* class_inited 	*/  FALSE,
/* initialize   	*/  initialize,
/* initialize_hook 	*/  NULL,
/* realize      	*/  XtInheritRealize,
/* actions      	*/  NULL,
/* num_actions  	*/  0,
/* resources    	*/  resources,
/* num_resources 	*/  5,
/* xrm_class    	*/  NULLQUARK,
/* compres_motion 	*/  False ,
/* compress_exposure 	*/  FALSE ,
/* compress_enterleave 	*/  False ,
/* visible_interest 	*/  False ,
/* destroy      	*/  destroy,
/* resize       	*/  XtInheritResize,
/* expose       	*/  XtInheritExpose,
/* set_values   	*/  set_values,
/* set_values_hook 	*/  NULL,
/* set_values_almost 	*/  XtInheritSetValuesAlmost,
/* get_values+hook 	*/  NULL,
/* accept_focus 	*/  XtInheritAcceptFocus,
/* version      	*/  XtVersion,
/* callback_private 	*/  NULL,
/* tm_table      	*/  NULL,
/* query_geometry 	*/  XtInheritQueryGeometry,
/* display_acceleator 	*/  XtInheritDisplayAccelerator,
/* extension    	*/  NULL 
},
{ /* composite_class part */
XtInheritGeometryManager,
XtInheritChangeManaged,
XtInheritInsertChild,
XtInheritDeleteChild,
NULL
},
{ /* Shell_class part */
 /* dummy */  0
},
{ /* WMShell_class part */
 /* dummy */  0
},
{ /* VendorShell_class part */
 /* dummy */  0
},
{ /* TransientShell_class part */
 /* dummy */  0
},
{ /* XfwfAlert_class part */
 /* dummy */  0
},
};
WidgetClass xfwfAlertWidgetClass = (WidgetClass) &xfwfAlertClassRec;
static void _resolve_inheritance(class)
WidgetClass class;
{
  XfwfAlertWidgetClass c = (XfwfAlertWidgetClass) class;
  XfwfAlertWidgetClass super;
  static CompositeClassExtensionRec extension_rec = {
    NULL, NULLQUARK, XtCompositeExtensionVersion,
    sizeof(CompositeClassExtensionRec), True};
  CompositeClassExtensionRec *ext;
  ext = (XtPointer)XtMalloc(sizeof(*ext));
  *ext = extension_rec;
  ext->next_extension = c->composite_class.extension;
  c->composite_class.extension = ext;
  if (class == xfwfAlertWidgetClass) return;
  super = (XfwfAlertWidgetClass)class->core_class.superclass;
}
#line 290 "Alert.w"
/*ARGSUSED*/
#if NeedFunctionPrototypes
#line 290 "Alert.w"
static void class_initialize(void)
#else
#line 290 "Alert.w"
static void class_initialize()
#endif
#line 291 "Alert.w"
{
    static XtConvertArgRec args[] = {
    { XtWidgetBaseOffset, 0, sizeof(Widget) } };

    XtSetTypeConverter(XtRString, "Icon", cvtStringToIcon,
		       args, XtNumber(args), XtCacheNone, NULL);
}
#line 303 "Alert.w"
/*ARGSUSED*/
#if NeedFunctionPrototypes
#line 303 "Alert.w"
static void initialize(Widget  request,Widget self,ArgList  args,Cardinal * num_args)
#else
#line 303 "Alert.w"
static void initialize(request,self,args,num_args)Widget  request;Widget self;ArgList  args;Cardinal * num_args;
#endif
#line 304 "Alert.w"
{
    ((XfwfAlertWidget)self)->xfwfAlert.outer = NULL;
    ((XfwfAlertWidget)self)->xfwfAlert.buttonw = NULL;
    ((XfwfAlertWidget)self)->xfwfAlert.labelw = NULL;
    ((XfwfAlertWidget)self)->xfwfAlert.iconw = NULL;
    ((XfwfAlertWidget)self)->xfwfAlert.num_buttons = NULL;
    ((XfwfAlertWidget)self)->xfwfAlert.buttons = XtNewString(((XfwfAlertWidget)self)->xfwfAlert.buttons);
    ((XfwfAlertWidget)self)->xfwfAlert.message = XtNewString(((XfwfAlertWidget)self)->xfwfAlert.message);
    ((XfwfAlertWidget)self)->xfwfAlert.outer = XtVaCreateManagedWidget("outer", xfwfBoardWidgetClass, self, NULL);
    create_children(self);
}
#line 319 "Alert.w"
/*ARGSUSED*/
#if NeedFunctionPrototypes
#line 319 "Alert.w"
static Boolean  set_values(Widget  old,Widget  request,Widget self,ArgList  args,Cardinal * num_args)
#else
#line 319 "Alert.w"
static Boolean  set_values(old,request,self,args,num_args)Widget  old;Widget  request;Widget self;ArgList  args;Cardinal * num_args;
#endif
#line 320 "Alert.w"
{
    Boolean recreate = False;

    if (((XfwfAlertWidget)old)->xfwfAlert.message != ((XfwfAlertWidget)self)->xfwfAlert.message) {
	XtFree(((XfwfAlertWidget)old)->xfwfAlert.message);
	((XfwfAlertWidget)self)->xfwfAlert.message = XtNewString(((XfwfAlertWidget)self)->xfwfAlert.message);
	recreate = True;
    }
    if (((XfwfAlertWidget)old)->xfwfAlert.buttons != ((XfwfAlertWidget)self)->xfwfAlert.buttons) {
	XtFree(((XfwfAlertWidget)old)->xfwfAlert.buttons);
	((XfwfAlertWidget)self)->xfwfAlert.buttons = XtNewString(((XfwfAlertWidget)self)->xfwfAlert.buttons);
	recreate = True;
    }
    if (((XfwfAlertWidget)old)->xfwfAlert.icon != ((XfwfAlertWidget)self)->xfwfAlert.icon || recreate) {
	create_children(self);
    }
    return False;
}
#line 339 "Alert.w"
/*ARGSUSED*/
#if NeedFunctionPrototypes
#line 339 "Alert.w"
static void destroy(Widget self)
#else
#line 339 "Alert.w"
static void destroy(self)Widget self;
#endif
#line 340 "Alert.w"
{
    XtFree(((XfwfAlertWidget)self)->xfwfAlert.buttons);
    XtFree(((XfwfAlertWidget)self)->xfwfAlert.message);
    XtFree((String) ((XfwfAlertWidget)self)->xfwfAlert.buttonw);
}
#line 118 "Alert.w"
#line 119 "Alert.w"
#line 121 "Alert.w"
/*ARGSUSED*/
#if NeedFunctionPrototypes
#line 121 "Alert.w"
int  XfwfDialog(String  name,Widget  parent,String  icon,String  msg,String  buttons)
#else
#line 121 "Alert.w"
int  XfwfDialog(name,parent,icon,msg,buttons)String  name;Widget  parent;String  icon;String  msg;String  buttons;
#endif
#line 122 "Alert.w"
{
    Widget alert;
    int choice = -1;
    XEvent event;
    Position x, y;
    Dimension wd, ht, wd1, ht1;

    x = ((XfwfAlertWidget)parent)->core.x;
    y = ((XfwfAlertWidget)parent)->core.y;
    wd = ((XfwfAlertWidget)parent)->core.width;
    ht = ((XfwfAlertWidget)parent)->core.height;
    alert = XtVaCreatePopupShell
	(name, xfwfAlertWidgetClass, parent,
	 XtVaTypedArg, XtNicon, XtRString, icon, strlen(icon) + 1,
	 XtNmessage, msg,
	 XtNbuttons, buttons,
	 NULL);
    XtAddCallback(alert, XtNcallback, button_pressed, &choice);
    XtRealizeWidget(alert);
    wd1 = ((XfwfAlertWidget)alert)->core.width;
    ht1 = ((XfwfAlertWidget)alert)->core.height;
    x = max(0, x + wd/2 - wd1/2);
    y = max(0, y + ht/2 - ht1/2);
    XtMoveWidget(alert, x, y);
    XtPopup(alert, XtGrabExclusive);
    while (choice == -1)
	XtAppProcessEvent(XtWidgetToApplicationContext(parent),
			  XtIMXEvent | XtIMTimer);
    XtDestroyWidget(alert);
    return choice;
}
